### PR TITLE
Add support for proactive messages

### PR DESCRIPTION
**Pull request by PrasadG193**
This PR is dependant on #26 and has to be merged after #26

### CODE REVIEW

File: connector/client/client.go [hunk 1]
```diff
@@ -84,18 +83,30 @@ func (client *ConnectorClient) Post(target url.URL, activity schema.Activity) er
...
+}
+
+func (client *ConnectorClient) checkRespError(resp *http.Response, err error) error {
+	allowedResp := []int{http.StatusOK, http.StatusCreated, http.StatusAccepted}
+	if err != nil {
 		return customerror.HTTPError{
-			StatusCode: resp.StatusCode,
-			HtErr:      err,
-			Body:       resp.Body,
+			HtErr: err,
```
Comment (sanketsudake) [inline connector/client/client.go]: Response object would be empty in this case ? Should we embed response attributes here as well as do below ?

File: connector/client/client.go [hunk 2]
```diff
@@ -84,18 +83,30 @@ func (client *ConnectorClient) Post(target url.URL, activity schema.Activity) er
...
+	for _, code := range allowedResp {
+		if code == resp.StatusCode {
+			return nil
+		}
+	}
+
+	return customerror.HTTPError{
+		HtErr:      errors.New("invalid response"),
+		StatusCode: resp.StatusCode,
+		Body:       resp.Body,
```
Comment (sanketsudake) [inline connector/client/client.go]: Body if too large we might end up embedding too much content in HTTPError.

File: core/activity/response.go [hunk 3]
```diff
@@ -20,21 +20,24 @@
...
 )
 
 // Response provides functionalities to send activity to the connector service.
 type Response interface {
 	SendActivity(activity schema.Activity) error
 }
 
-const replyToAcitivityURL = "v3/conversations/{conversationId}/activities/{activityId}"
+const (
+	sendToConversationURL = "/v3/conversations/%s/activities"
```
Comment (sanketsudake) [inline core/activity/response.go]: Please see how we can make version upgrades seemless. Eg. If we need to update SDK to v4 API.
Probably we can define constant  for MS Teams API version we  are using and change it globally whenever we are shifting API compatibility globally.

File: core/activity/response.go [hunk 4]
```diff
@@ -43,50 +46,22 @@ type DefaultResponse struct {
...
+	// if ReplyToID is set in the activity, we send reply to that particular activity
+	if activity.ReplyToID != "" {
+		respPath = fmt.Sprintf(replyToActivityURL, activity.Conversation.ID, activity.ID)
+	}
 
-	return *u, nil
+	// Send activity to client
+	u.Path = path.Join(u.Path, respPath)
+	err = response.Client.Post(*u, activity)
+	return errors.Wrap(err, "Failed to send response.")
```
Comment (sanketsudake) [inline core/activity/response.go]: Just confirming my assumption. Wrap would ignore nil error.

File: schema/customerror/HttpError.go [hunk 5]
```diff
@@ -33,7 +33,11 @@ type HTTPError struct {
 }
 
 func (htErr HTTPError) Error() string {
-	buf := new(bytes.Buffer)
-	buf.ReadFrom(htErr.Body)
-	return fmt.Sprintf("HTTP error %d: %s %s", htErr.StatusCode, htErr.HtErr, buf.String())
+	msg := fmt.Sprintf("HTTP error %d: %s.", htErr.StatusCode, htErr.HtErr)
+	if htErr.Body != nil {
```
Comment (sanketsudake) [inline schema/customerror/HttpError.go]: If body is too large. Please check what would be its impact on err string.

File: core/bot_framework_adapter.go [hunk 6]
```diff
@@ -92,12 +93,32 @@ func (bf *BotFrameworkAdapter) ProcessActivity(ctx context.Context, req schema.A
...
+
+	replyActivity, err := activity.PrepareActivityContext(handler, turnContext)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create Activity context.")
 	}
 
 	response, err := activity.NewActivityResponse(bf.Client)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Failed to create response object.")
```
Comment (sahil-lakhwani) [inline core/bot_framework_adapter.go]: @PrasadG193 ProcessActivity and ProactiveMessage function have some same code. Can we take that code out?

**Comment by sahil-lakhwani** [comment]
Let's add an example for proactive messages.

File: core/activity/response.go (hunk 4 above)
Comment (PrasadG193) [reply core/activity/response.go]: Yes. It will return nil

File: core/bot_framework_adapter.go (hunk 6 above)
Comment (PrasadG193) [reply core/bot_framework_adapter.go]: Sounds good!

**Comment by PrasadG193** [comment]
@sahil-lakhwani I'll raise a separate PR for the example.

File: connector/client/client.go (hunk 1 above)
Comment (PrasadG193) [reply connector/client/client.go]: response would be nil in this case

File: connector/client/client.go (hunk 2 above)
Comment (PrasadG193) [reply connector/client/client.go]: Make sense. Let's skip embedding response body

File: schema/customerror/HttpError.go (hunk 5 above)
Comment (PrasadG193) [reply schema/customerror/HttpError.go]: We are not using this internally anymore.

Comment (sanketsudake) [reply schema/customerror/HttpError.go]: This is still pending ? Body I assume will be nil with new changes.

**Pull request merged**